hack/generator/pkg/astmodel/type_walker.go

```diff
@@ -0,0 +1,194 @@
+// TODO: but unsure if there's a better way... bring it up in code review.
+
+// TypeWalker performs a depth first traversal across the types provided, applying the visitor to each TypeDefinition.
+// MakeContext is called before each visit, and AfterVisit is called after each visit. RemoveCycle is called
+// if a cycle is detected.
+type TypeWalker struct {
+	allTypes Types
+	visitor  TypeVisitor
+
+	// MakeContext is called before any type but the root type is visited. It is given the current context and returns
+	// a new context for use in the upcoming Visit.
+	MakeContext func(it TypeName, ctx interface{}) (interface{}, error)
+	// AfterVisit is called after the type walker has applied the visitor to a TypeDefinition.
+	AfterVisit func(original TypeDefinition, updated TypeDefinition, ctx interface{}) (TypeDefinition, error)
+	// RemoveCycle is called if a cycle is detected. If true is returned the cycle will be pruned, otherwise it will be preserved as-is.
+	RemoveCycle func(def TypeDefinition, ctx interface{}) (bool, error)
+
+	state                   typeWalkerState
+	originalVisitTypeName   func(this *TypeVisitor, it TypeName, ctx interface{}) (Type, error)
+	originalVisitObjectType func(this *TypeVisitor, it *ObjectType, ctx interface{}) (Type, error)
+}
+
+type typeWalkerState struct {
+	result     Types
+	processing map[TypeName]struct{}
+}
+
+// NewTypeWalker returns a TypeWalker.
+// The provided visitor VisitTypeName function must return a TypeName and VisitObjectType must return an ObjectType or calls to
+// Walk will panic.
+func NewTypeWalker(allTypes Types, visitor TypeVisitor) *TypeWalker {
+	typeWalker := TypeWalker{
+		allTypes:                allTypes,
+		originalVisitTypeName:   visitor.VisitTypeName,
+		originalVisitObjectType: visitor.VisitObjectType,
+		AfterVisit:              DefaultAfterVisit,
+		MakeContext:             IdentityMakeContext,
+		RemoveCycle:             IdentityRemoveCycle,
+	}
+
+	// visitor is a copy - modifications won't impact passed visitor
+	visitor.VisitTypeName = typeWalker.visitTypeName
+	visitor.VisitObjectType = typeWalker.visitObjectType
+
+	typeWalker.visitor = visitor
+
+	return &typeWalker
+}
+
+func (t *TypeWalker) visitTypeName(this *TypeVisitor, it TypeName, ctx interface{}) (Type, error) {
+	updatedCtx, err := t.MakeContext(it, ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	visitedTypeName, err := t.originalVisitTypeName(this, it, updatedCtx)
+	if err != nil {
+		return nil, err
+	}
+	it, ok := visitedTypeName.(TypeName)
+	if !ok {
+		panic(fmt.Sprintf("TypeWalker visitor VisitTypeName must return a TypeName, instead returned %T", visitedTypeName))
+	}
+
+	def, ok := t.allTypes[it]
+	if !ok {
+		return nil, errors.Errorf("couldn't find type %q", it)
+	}
+
+	// Prevent loops by bypassing this type if it's currently being processed.
+	if _, ok := t.state.processing[def.Name()]; ok {
+		remove, err := t.RemoveCycle(def, updatedCtx)
```

Comment:
Does it matter that this can be called with any element of the cycle?

'cause I don't believe this is deterministic. The `TypeVisitor`' method `IdentityVisitOfObjectType` iterates over `properties` without sorting (it's a map), so a cycle might be entered at a different point from run to run.

To illustrate:

``` go
type Person struct {
    enrolments []Enrolment
}

type Enrolment {
    student Person
    course  Course
}

type Course struct {
    lecturer Person
    tutors   []Person
    students []Enrolment
}
```
One cycle is:  
Person.enrolments -> Enrolment.student -> Person

Another is
Person.enrolments -> Enrolment.course -> Course.lecturer -> Person

Another is
Person.enrolments -> Enrolment.course -> Course.students -> Enrolment

Depending on which cycle is encountered first, won't you end up with different results from the walk?

Reply:
Wow really good @theunrepentantgeek.

Yes I found and fixed this bug in my future PR. I will backport it to this PR as you point it it can cause problems here.
I specifically wrote a golden file test for this case (which will stay in my upcoming PR) because as you say, it can happen!

Really impressed you caught that just with a quick review, it took me a while to figure out what was happening when I ran into it!

Reply:
Pushed my fix for this (which makes `IdentityVisitOfObjectType` iterate properties in a deterministic order).

Reply:
@matthchr it's mostly a case of damaged shins. I've walked into the furniture of non-deterministic execution enough times that I've got (metaphorical) scars.